ops/framework.py

```diff
@@ -734,6 +755,23 @@ def _reemit(self, single_event_path=None):
         if not deferred:
             self._storage.drop_snapshot(last_event_path)
 
+    def breakpoint(self, name=None):
+        """Provide a debugger breakpoint if proper execution context."""
+        if name is not None:
+            # validate the given name
+            if not isinstance(name, str):
+                raise TypeError('breakpoint names must be strings')
+            if name in {JUJU_DEBUG_HOOK, JUJU_DEBUG_ALL}:
+                raise ValueError('breakpoint names "all" and "hook" are forbidden')
+            if not re.match(r'^[a-z0-9]([a-z0-9\-]*[a-z0-9])?$', name):
```

Comment:
Regexes are cached. So this will be compiled the first used time, and then used the compilated one, automatically. The cost is a dict lookup in regex cache, nothing else.

If we pre-compile this at module level, will be paying the compilation cost every time the process is executed, even if `.breakpoint()` is not used at all (which I expect to be the majority of cases).